ic-cdk/src/call.rs

```diff
@@ -1,105 +1,84 @@
+use thiserror::Error;
 
 pub use ic_response_codes::RejectCode;
 
-const CALL_PERFORM_REJECT_MESSAGE: &str = "call_perform failed";
-
-/// The error type for inter-canister calls and decoding the response.
-///
-/// See [`Call::call`] or [`Call::call_tuple`].
-#[derive(thiserror::Error, Debug, Clone)]
-pub enum CallError {
-    /// The call was rejected.
-    ///
-    /// Please handle the error by matching on the rejection code.
-    #[error("The call was rejected with code {0:?}")]
-    CallRejected(CallRejected),
-
-    /// The response could not be decoded.
-    #[error("Failed to decode the response as {0}")]
-    CandidDecodeFailed(String),
-}
-
-/// The error type for inter-canister calls.
-///
-/// See [`Call::call_raw`] and [`Call::call_oneway`].
-///
-/// This type is also the inner error type of [`CallError::CallRejected`].
-#[derive(Debug, Clone)]
-pub struct CallRejected {
-    /// All fields are private so we will be able to change the implementation without breaking the API.
-    /// Once we have `ic0.msg_error_code` system API, we will only store the `error_code` in this struct.
-    /// It will still be possible to get the [`RejectCode`] using the public getter,
-    /// because every `error_code` can map to a [`RejectCode`].
-    reject_code: RejectCode,
-    reject_message: String,
-    sync: bool,
-}
-
-impl CallRejected {
-    /// Gets the [`RejectCode`].
-    pub fn reject_code(&self) -> RejectCode {
-        self.reject_code
-    }
-
-    /// Retrieves the reject message associated with the call.
-    ///
-    /// - For an asynchronous rejection (when the IC rejects the call after it was enqueued),
-    ///   the message is obtained from [`api::msg_reject_msg`](`msg_reject_msg`).
-    /// - For a synchronous rejection (when `ic0.call_perform` returns a non-zero code),
-    ///   the message is set to a fixed string: `"call_perform failed"`.
-    pub fn reject_message(&self) -> &str {
-        &self.reject_message
-    }
-
-    /// Checks if the call was rejected synchronously or asynchronously.
-    ///
-    /// A synchronous rejection occurs when `ic0.call_perform` returns a non-zero code immediately.
-    /// An asynchronous rejection happens when the call is enqueued but later rejected by the IC.
-    ///
-    /// # Returns
-    /// - `true` if the call was rejected synchronously.
-    /// - `false` if the call was rejected asynchronously.
-    pub fn is_sync(&self) -> bool {
-        self.sync
-    }
-}
-
-/// Result of a inter-canister call.
-pub type SystemResult<R> = Result<R, CallRejected>;
-
-/// Result of a inter-canister call and decoding the response.
-pub type CallResult<R> = Result<R, CallError>;
-
 /// Inter-canister Call.
 ///
 /// This type enables the configuration and execution of inter-canister calls using a builder pattern.
 ///
+/// # Constructors
+///
+/// [`Call`] has two constructors that differentiate whether the call is made unboundedly waiting for a response or not.
```

Comment:
Maybe a bit unclear at first glance?
```suggestion
/// [`Call`] has two constructors that differentiate whether the call's response is waited for an unbounded amount of time or not.
```